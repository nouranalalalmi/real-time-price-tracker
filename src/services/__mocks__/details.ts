export const useGetAssetDetails = jest.fn(() => ({
  data: {
    id: 'bitcoin',
    rank: '1',
    symbol: 'BTC',
    name: 'Bitcoin',
    supply: '19693006.0000000000000000',
    maxSupply: '21000000.0000000000000000',
    marketCapUsd: '1154471549507.9040036854136828',
    volumeUsd24Hr: '13866421449.4822673574551073',
    priceUsd: '58623.4295316775917138',
    changePercent24Hr: '2.2406824871311752',
    vwap24Hr: '57798.6225887850445657',
    explorer: 'https://blockchain.info/',
  },
  isLoading: false,
}));

export const useGetAssetHistory = jest.fn(() => ({
  data: [
    {
      priceUsd: '57453.4654235888369902',
      time: 1714575600000,
      circulatingSupply: '19692650.0000000000000000',
      date: '2024-05-01T15:00:00.000Z',
    },
    {
      priceUsd: '56849.8205946134484453',
      time: 1714579200000,
      circulatingSupply: '19692650.0000000000000000',
      date: '2024-05-01T16:00:00.000Z',
    },
    {
      priceUsd: '56989.2744984085723538',
      time: 1714582800000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T17:00:00.000Z',
    },
    {
      priceUsd: '57421.7196556879817478',
      time: 1714586400000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T18:00:00.000Z',
    },
    {
      priceUsd: '58915.9403189355241213',
      time: 1714590000000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T19:00:00.000Z',
    },
    {
      priceUsd: '57038.2007750086044715',
      time: 1714593600000,
      circulatingSupply: '19692693.0000000000000000',
      date: '2024-05-01T20:00:00.000Z',
    },
    {
      priceUsd: '57342.4529782201663178',
      time: 1714597200000,
      circulatingSupply: '19692737.0000000000000000',
      date: '2024-05-01T21:00:00.000Z',
    },
    {
      priceUsd: '57871.1674520177401757',
      time: 1714600800000,
      circulatingSupply: '19692737.0000000000000000',
      date: '2024-05-01T22:00:00.000Z',
    },
    {
      priceUsd: '57904.7652767478518852',
      time: 1714604400000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-01T23:00:00.000Z',
    },
    {
      priceUsd: '58350.5875656442168466',
      time: 1714608000000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T00:00:00.000Z',
    },
    {
      priceUsd: '57759.0173401927487826',
      time: 1714611600000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T01:00:00.000Z',
    },
    {
      priceUsd: '57173.5786733097720706',
      time: 1714615200000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T02:00:00.000Z',
    },
    {
      priceUsd: '57451.8361411585783057',
      time: 1714618800000,
      circulatingSupply: '19692793.0000000000000000',
      date: '2024-05-02T03:00:00.000Z',
    },
    {
      priceUsd: '57495.4104137510215865',
      time: 1714622400000,
      circulatingSupply: '19692837.0000000000000000',
      date: '2024-05-02T04:00:00.000Z',
    },
    {
      priceUsd: '57362.3002691941893159',
      time: 1714626000000,
      circulatingSupply: '19692837.0000000000000000',
      date: '2024-05-02T05:00:00.000Z',
    },
    {
      priceUsd: '57462.2929728636015241',
      time: 1714629600000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T06:00:00.000Z',
    },
    {
      priceUsd: '57596.3996240546355440',
      time: 1714633200000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T07:00:00.000Z',
    },
    {
      priceUsd: '57669.7948503898871027',
      time: 1714636800000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T08:00:00.000Z',
    },
    {
      priceUsd: '57879.2448705878824941',
      time: 1714640400000,
      circulatingSupply: '19692887.0000000000000000',
      date: '2024-05-02T09:00:00.000Z',
    },
    {
      priceUsd: '57816.8318253461107454',
      time: 1714644000000,
      circulatingSupply: '19692946.0000000000000000',
      date: '2024-05-02T10:00:00.000Z',
    },
    {
      priceUsd: '57821.9173267937048981',
      time: 1714647600000,
      circulatingSupply: '19692946.0000000000000000',
      date: '2024-05-02T11:00:00.000Z',
    },
    {
      priceUsd: '58537.5857861513969580',
      time: 1714651200000,
      circulatingSupply: '19693006.0000000000000000',
      date: '2024-05-02T12:00:00.000Z',
    },
    {
      priceUsd: '58634.5239917690989039',
      time: 1714654800000,
      circulatingSupply: '19693006.0000000000000000',
      date: '2024-05-02T13:00:00.000Z',
    },
    {
      priceUsd: '58479.7446811632706652',
      time: 1714658400000,
      circulatingSupply: '19693006.0000000000000000',
      date: '2024-05-02T14:00:00.000Z',
    },
  ],
  isLoading: false,
}));
