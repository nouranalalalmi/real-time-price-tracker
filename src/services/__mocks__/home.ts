export const useGetAssetPrice = jest.fn(() => ({
  data: [
    {
      symbol: 'BTCUSDT',
      priceChange: '1154.47000000',
      priceChangePercent: '2.010',
      weightedAvgPrice: '57925.32781024',
      prevClosePrice: '57423.91000000',
      lastPrice: '58578.39000000',
      lastQty: '0.00040000',
      bidPrice: '58578.39000000',
      bidQty: '1.80801000',
      askPrice: '58578.40000000',
      askQty: '2.46959000',
      openPrice: '57423.92000000',
      highPrice: '59468.00000000',
      lowPrice: '56625.09000000',
      volume: '64977.59196000',
      quoteVolume: '3763848314.60286340',
      openTime: 1714573846188,
      closeTime: 1714660246188,
      firstId: 3581957071,
      lastId: 3583882631,
      count: 1925561,
    },
    {
      symbol: 'ETHUSDT',
      priceChange: '77.99000000',
      priceChangePercent: '2.695',
      weightedAvgPrice: '2949.08665918',
      prevClosePrice: '2894.00000000',
      lastPrice: '2971.99000000',
      lastQty: '1.34440000',
      bidPrice: '2971.99000000',
      bidQty: '3.65560000',
      askPrice: '2972.00000000',
      askQty: '35.06470000',
      openPrice: '2894.00000000',
      highPrice: '3015.73000000',
      lowPrice: '2865.91000000',
      volume: '500554.19830000',
      quoteVolume: '1476177708.40407100',
      openTime: 1714573846184,
      closeTime: 1714660246184,
      firstId: 1411435856,
      lastId: 1412548706,
      count: 1112851,
    },
    {
      symbol: 'ADAUSDT',
      priceChange: '0.01690000',
      priceChangePercent: '3.866',
      weightedAvgPrice: '0.44709554',
      prevClosePrice: '0.43720000',
      lastPrice: '0.45400000',
      lastQty: '1189.70000000',
      bidPrice: '0.45400000',
      bidQty: '19898.00000000',
      askPrice: '0.45410000',
      askQty: '3031.00000000',
      openPrice: '0.43710000',
      highPrice: '0.46250000',
      lowPrice: '0.42900000',
      volume: '122243638.70000000',
      quoteVolume: '54654585.17339000',
      openTime: 1714573845685,
      closeTime: 1714660245685,
      firstId: 488692403,
      lastId: 488807792,
      count: 115390,
    },
    {
      symbol: 'SOLUSDT',
      priceChange: '13.67000000',
      priceChangePercent: '11.136',
      weightedAvgPrice: '131.78729373',
      prevClosePrice: '122.76000000',
      lastPrice: '136.43000000',
      lastQty: '54.16200000',
      bidPrice: '136.42000000',
      bidQty: '233.29900000',
      askPrice: '136.43000000',
      askQty: '18.50600000',
      openPrice: '122.76000000',
      highPrice: '139.20000000',
      lowPrice: '121.41000000',
      volume: '7684834.78500000',
      quoteVolume: '1012763579.10641000',
      openTime: 1714573846187,
      closeTime: 1714660246187,
      firstId: 551238525,
      lastId: 552376984,
      count: 1138460,
    },
  ],
  isLoading: false,
}));
