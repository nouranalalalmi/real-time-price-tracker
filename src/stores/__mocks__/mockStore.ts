export const mockStore = (store: string) => {
  jest.doMock(store, () => {
    const actualObject = jest.requireActual(store);
    return {
      ...actualObject,
      getState: jest.fn(),
    };
  });
};
