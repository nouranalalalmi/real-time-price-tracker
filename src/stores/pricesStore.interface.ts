type Price = {
  [key: string]: {
    lastPrice: string;
    priceChangePercent: string;
  };
};

export interface PricesStoreInterface {
  prices: Price;
  setPrices: (price: Price) => void;
}
