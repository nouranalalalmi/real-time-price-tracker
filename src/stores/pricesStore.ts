import { create } from 'zustand';
import { PricesStoreInterface } from './pricesStore.interface';

export const usePriceStore = create<PricesStoreInterface>(set => ({
  prices: {},
  setPrices: newPrices =>
    set(state => ({
      prices: {
        ...state.prices,
        ...newPrices,
      },
    })),
}));
