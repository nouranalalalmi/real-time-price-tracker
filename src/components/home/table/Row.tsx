import Image from 'next/image';
import { useRouter } from 'next/navigation';

import { usePriceStore } from '@/stores/pricesStore';
import { FormattedNumber } from '@/ui/formattedNumber/FormattedNumber';
import { PriceChangeIndicator } from '@/ui/priceChangeIndicator/PriceChangeIndicator';

import { AssetPriceInterface } from './Row.interface';

export const Row = ({ asset }: AssetPriceInterface) => {
  const prices = usePriceStore(state => state.prices);
  const price = prices[asset.tickerId.toUpperCase()];
  const router = useRouter();

  return (
    <tr
      className="cursor-pointer hover:bg-gray-100"
      onClick={() => router.push(`/details/${asset.id}`)}>
      <td className="flex items-center gap-x-2">
        <div className="relative size-9">
          <Image
            src={asset.logo}
            alt={asset.name}
            layout="fill"
            objectFit="contain"
            objectPosition="center"
          />
        </div>
        <div className="flex flex-col">
          <span className="text-lg uppercase">{asset.symbol}</span>
          <span className="text-xs text-gray-500">{asset.name}</span>
        </div>
      </td>
      <td>
        <FormattedNumber prefix="$" value={price?.lastPrice} />
      </td>
      <td>
        <FormattedNumber prefix="$" value={price?.priceChange} />
      </td>
      <td>
        <PriceChangeIndicator percent={price?.priceChangePercent} />
      </td>
    </tr>
  );
};
