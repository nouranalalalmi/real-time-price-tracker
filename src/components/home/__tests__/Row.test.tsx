import { render, screen } from '@testing-library/react';
import mockRouter from 'next-router-mock';

import { mockStore } from '@/stores/__mocks__/mockStore';
import { usePriceStore } from '@/stores/pricesStore';

import { Row } from '../Row';

mockStore('@/stores/pricesStore');

describe('Row', () => {
  const mockAsset = {
    id: 'bitcoin',
    tickerId: 'btcusdt',
    logo: '/logo.png',
    name: 'Bitcoin',
    symbol: 'BTC',
  };

  beforeEach(() => {
    usePriceStore.setState({
      prices: {
        BTCUSDT: {
          lastPrice: '50000',
          priceChange: '250',
          priceChangePercent: '5',
        },
      },
    });
  });

  it('should match snapshot', () => {
    const { container } = render(<Row asset={mockAsset} />);
    expect(container).toMatchSnapshot();
  });

  it('should display asset information and price', () => {
    render(<Row asset={mockAsset} />);

    expect(screen.getByText('BTC')).toBeInTheDocument();
    expect(screen.getByText('Bitcoin')).toBeInTheDocument();
    expect(screen.getByText('$250.00')).toBeInTheDocument();
    expect(screen.getByText('$50,000.00')).toBeInTheDocument();
    expect(screen.getByText('5.00%')).toBeInTheDocument();
  });

  it('should navigate to asset page when row is clicked', () => {
    render(<Row asset={mockAsset} />);

    screen.getByText('BTC').click();
    expect(mockRouter).toMatchObject({ pathname: '/details/bitcoin' });
  });
});
