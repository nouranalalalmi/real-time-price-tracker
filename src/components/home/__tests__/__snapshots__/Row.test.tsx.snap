// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row should match snapshot 1`] = `
<div>
  <tr
    class="cursor-pointer hover:bg-gray-100"
  >
    <td
      class="flex items-center gap-x-2"
    >
      <div
        class="relative size-9"
      >
        <img
          alt="Bitcoin"
          src="/logo.png"
        />
      </div>
      <div
        class="flex flex-col"
      >
        <span
          class="text-lg uppercase"
        >
          BTC
        </span>
        <span
          class="text-xs text-gray-500"
        >
          Bitcoin
        </span>
      </div>
    </td>
    <td>
      $
      50,000.00
    </td>
    <td>
      $
      250.00
    </td>
    <td>
      <div
        class="bg-green-100 text-green-500 w-fit rounded-xl px-2 py-1 text-xs font-semibold"
      >
        5.00
        %
      </div>
    </td>
  </tr>
</div>
`;
